Python 3.8 as minimum

mr-c opened the pull request:
![Walrus](https://media.tenor.com/McSwVUFwyV8AAAAC/50first-dates-jocko-the-walrus.gif "happy walrus")

Fixes: https://github.com/galaxyproject/galaxy/issues/16394

## Major or minor packages bumps now that Python 3.8 is the minimum
- a2wsgi [1.7.0 → 1.8.0](https://github.com/abersheeran/a2wsgi/compare/v1.7.0...v1.8.0)
- bleach [6.0.0 → 6.1.0](https://bleach.readthedocs.io/en/latest/changes.html)
- celery [5.2.7 → 5.3.6](https://github.com/celery/celery/blob/main/Changelog.rst)
- cwltool [3.1.20221109155812 → 3.1.20231114134824](https://github.com/common-workflow-language/cwltool/releases)
  - job loading: allow inputs named `id`
  - Fixed race condition which happens when a job runs "too fast"
  - Properly locate podman images 
  - Python 3.12 support
  - required outputs with `secondaryFiles` fix for containers
  - pass through `QEMU_LD_PREFIX` by default
  - Ensure that files with identical basenames get staged to unique paths
  - always print floats on the command line with decimal notation
  - CWL expressions in `ResourceRequirement` fields can return floats in CWL v1.2+
  - support podman even when not installed as "docker" alias
- dnspython [2.3.0 → 2.4.2](https://dnspython.readthedocs.io/en/latest/whatsnew.html)
- h5py [3.8.0 → 3.10.0](https://docs.h5py.org/en/stable/whatsnew/index.html)
- importlib-resources [5.12.0 → 6.1.1 (backports from Python 3.13)](https://github.com/python/importlib_resources/blob/main/NEWS.rst) [see also](https://github.com/python/importlib_resources#compatibility)
- kombu [5.2.4 → 5.3.4](https://docs.celeryq.dev/projects/kombu/en/stable/changelog.html)
- markdown [3.4.4 → 3.5.1](https://python-markdown.github.io/changelog/#35-2023-10-06)
- numpy [1.21.6 → 1.24.4+](https://numpy.org/doc/stable/release/1.22.0-notes.html)
- pyparsing [3.0.9 → 3.1.1](https://github.com/pyparsing/pyparsing/blob/master/CHANGES)
- tifffile [2021.11.2 → 2023.7.10](https://github.com/cgohlke/tifffile/blob/master/CHANGES.rst)
- uvicorn [0.22.0 → 0.24.0.post1](https://github.com/encode/uvicorn/blob/master/CHANGELOG.md)
- uvloop [0.18.0 → 0.19.0](https://github.com/MagicStack/uvloop/releases)
### Lint
- flake8 [5.0.4 → 6.1.0 (Improve multiprocessing performance)](https://flake8.pycqa.org/en/latest/release-notes/index.html)
- pycodestyle [2.9.1 → 2.11.1](https://pycodestyle.pycqa.org/en/latest/developer.html#changes)
- pyflakes [2.5.0 → 3.1.0](https://github.com/PyCQA/pyflakes/blob/main/NEWS.rst)
### Dev
- black [23.3.0 → 23.11.0](https://github.com/psf/black/blob/main/CHANGES.md#23110)
- cwltest [2.3.20230825125225 → 2.4.20231031143732 (favor "path" over "location" when testing, for @mvdbeek )](https://github.com/common-workflow-language/cwltest/releases)
- httpx [0.24.1 → 0.25.2](https://github.com/encode/httpx/blob/master/CHANGELOG.md)
- myst-parser [1.0.0 → 2.0.0](https://myst-parser.readthedocs.io/en/latest/develop/_changelog.html)
- playwright [1.35.0 → 1.40.0](https://playwright.dev/python/docs/release-notes)
- pytest-html [3.2.0 → 4.1.1](https://pytest-html.readthedocs.io/en/latest/changelog.html)
- pytest-postgresql [4.1.1 → 5.0.0 (Support Python 3.11)](https://github.com/ClearcodeHQ/pytest-postgresql/blob/main/CHANGES.rst)
- selenium [4.11.2 → 4.15.2](https://github.com/SeleniumHQ/selenium/blob/trunk/py/CHANGES)
- sphinx [5.3.0 → 7.1.2](https://www.sphinx-doc.org/en/master/changes.html)
### Typecheck
- mypy [1.4.1 → 1.7.1](https://github.com/python/mypy/blob/master/CHANGELOG.md#mypy-17) :partying_face: 

If any of the above release have desired features, we should document their new minimum versions (which will also speed up future `poetry` runs).

- [x] Can importlib-resources be removed?
- [x] Can importlib-metadata be removed?
- [x] [Bleach is deprecated since 2023-01-23](https://github.com/mozilla/bleach/issues/698) What to do? → https://github.com/galaxyproject/galaxy/issues/17073

## How to test the changes?
(Select all options that apply)
- ~~I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).~~
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mr-c committed:
Python 3.8 as new minimum version

Fixes: https://github.com/galaxyproject/galaxy/issues/16394

Also upgrade code to Python 3.8 syntax with:

```
ack --type=python -f | grep -v '^lib/galaxy/schema/bco/\|^lib/galaxy/schema/drs/\|lib/tool_shed_client/schema/trs.*.py\|^tools/\|^.venv/\|^.tox/' | grep -v '^lib/galaxy/files/sources/\|^lib/galaxy/job_metrics/\|^lib/galaxy/objectstore/\|^lib/galaxy/tool_util/\|^lib/galaxy/util/' | xargs pyupgrade --py38-plus
```

mr-c committed:
Use the walrus `:=` operator!

Fixed by running:

```
ack --type=python -f | grep -v '^lib/galaxy/schema/bco/\|^lib/galaxy/schema/drs/\|^tools/\|^.venv/\|^.tox/' | grep -v '^lib/galaxy/files/sources/\|^lib/galaxy/job_metrics/\|^lib/galaxy/objectstore/\|^lib/galaxy/tool_util/\|^lib/galaxy/util/' | xargs auto-walrus
make format
```

Note that the directories for the packages listed in
`packages/packages_for_pulsar_by_dep_dag.txt`
were explictly excluded.

mr-c committed:
need newer types-request for this class

mr-c committed:
typo fix

nsoranzo commented on .github/workflows/test_galaxy_packages_for_pulsar.yaml:
```diff
@@ -20,7 +20,7 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: ['3.7']
+        python-version: ['3.8']
```
Can you revert this please? We need to keep testing these packages on 3.7 (that's the whole point of this workflow, see details in #16394).

nsoranzo commented on test/unit/app/jobs/test_rule_helper.py:
```diff
@@ -56,9 +56,7 @@ def test_job_count():
-    if expected_count != acutal_count:
+    if expected_count != (acutal_count := rule_helper.job_count(**kwds)):
         template = "Expected job count %d, actual job count %s for params %s"
         raise AssertionError(template % (expected_count, acutal_count, kwds))
```
```suggestion
    if expected_count != (actual_count := rule_helper.job_count(**kwds)):
        raise AssertionError(f"Expected job count {expected_count}, actual job count {actual_count} for params {kwds}")
```

mr-c committed:
pyproject.toml: bump minimum cwltool version to save time

mr-c committed:
packages: synchronize pins with pyproject.toml

mr-c committed:
Poetry refresh

with `make update-requirements`

mr-c committed:
MyPy 1.7.0 tweaks

"Unpack is already enabled by default" in version 1.7.0

nsoranzo commented on lib/galaxy/util/commands.py:
```diff
@@ -68,9 +68,7 @@ def shell(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **k
-def shell_process(
-    cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: Dict[str, Any]
-) -> subprocess.Popen:
+def shell_process(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: str) -> subprocess.Popen:
```
```suggestion
def shell_process(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: Any) -> subprocess.Popen:
```

nsoranzo commented on lib/galaxy/util/commands.py:
```diff
@@ -55,7 +55,7 @@ def redirect_aware_commmunicate(p, sys=_sys):
 
 
-def shell(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: Dict[str, Any]) -> int:
+def shell(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: str) -> int:
```
For `**kwargs` the type annotation (pre Python 3.11) should be the type of the *values* (see e.g. https://peps.python.org/pep-0692/ ), which can be any in this case.

```suggestion
def shell(cmds: Union[List[str], str], env: Optional[Dict[str, str]] = None, **kwds: Any) -> int:
```

mr-c replied:
Thanks!

nsoranzo committed:
Pin twill to >=3.2.1

twill 3.2 replaced `requests` with `httpx`, and `httpx` has a default
timeout of 5 s on a request, while `requests` didn't enforce a timeout by
default.

Fix errors like the following in ToolShed tests, where resetting
repository metadata takes longer than 5 s:

```
self = <tool_shed.test.functional.test_0000_basic_repository_features.TestBasicRepositoryFeatures object at 0x7f28f99197c0>

    def test_0090_verify_repository_metadata(self):
        """Verify that resetting the metadata does not change it."""
        repository = self._get_repository_by_name_and_owner(repository_name, common.test_user_1_name)
>       self.verify_unchanged_repository_metadata(repository)

lib/tool_shed/test/functional/test_0000_basic_repository_features.py:264:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lib/tool_shed/test/base/twilltestcase.py:2073: in verify_unchanged_repository_metadata
    self.reset_repository_metadata(repository)
lib/tool_shed/test/base/twilltestcase.py:1774: in reset_repository_metadata
    self.visit_url("/repository/reset_all_metadata", params=params)
lib/tool_shed/test/base/twilltestcase.py:845: in visit_url
    return self._browser.visit_url(url, allowed_codes=allowed_codes)
lib/tool_shed/test/base/twillbrowser.py:35: in visit_url
    return visit_url(url, allowed_codes=allowed_codes)
lib/tool_shed/test/base/twillbrowser.py:20: in visit_url
    new_url = tc.go(url)
.venv/lib/python3.8/site-packages/twill/commands.py:120: in go
    browser.go(url)
.venv/lib/python3.8/site-packages/twill/browser.py:192: in go
    self._journey("open", try_url)
.venv/lib/python3.8/site-packages/twill/browser.py:677: in _journey
    result = self._client.get(url)
.venv/lib/python3.8/site-packages/httpx/_client.py:1041: in get
    return self.request(
.venv/lib/python3.8/site-packages/httpx/_client.py:814: in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
.venv/lib/python3.8/site-packages/httpx/_client.py:901: in send
    response = self._send_handling_auth(
.venv/lib/python3.8/site-packages/httpx/_client.py:929: in _send_handling_auth
    response = self._send_handling_redirects(
.venv/lib/python3.8/site-packages/httpx/_client.py:966: in _send_handling_redirects
    response = self._send_single_request(request)
.venv/lib/python3.8/site-packages/httpx/_client.py:1002: in _send_single_request
    response = transport.handle_request(request)
.venv/lib/python3.8/site-packages/httpx/_transports/default.py:228: in handle_request
    resp = self._pool.handle_request(req)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/contextlib.py:131: in __exit__
    self.gen.throw(type, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @contextlib.contextmanager
    def map_httpcore_exceptions() -> typing.Iterator[None]:
        try:
            yield
        except Exception as exc:
            mapped_exc = None

            for from_exc, to_exc in HTTPCORE_EXC_MAP.items():
                if not isinstance(exc, from_exc):
                    continue
                # We want to map to the most specific exception we can find.
                # Eg if `exc` is an `httpcore.ReadTimeout`, we want to map to
                # `httpx.ReadTimeout`, not just `httpx.TimeoutException`.
                if mapped_exc is None or issubclass(to_exc, mapped_exc):
                    mapped_exc = to_exc

            if mapped_exc is None:  # pragma: no cover
                raise

            message = str(exc)
>           raise mapped_exc(message) from exc
E           httpx.ReadTimeout: timed out

.venv/lib/python3.8/site-packages/httpx/_transports/default.py:83: ReadTimeout
```

nsoranzo commented on tox.ini:
```diff
@@ -39,6 +39,8 @@ passenv =
     GALAXY_CONFIG_OVERRIDE_DATABASE_CONNECTION
     APP_WEBSERVER
     TERM
+    TEST_PYTHON
+    TZ
```
Are these needed?

mr-c replied:
For local testing, yes. I needed `TZ` for a timezone synonym issue, and `TEST_PYTHON` to easily influence the version of Python tested

nsoranzo replied:
Doesn't `tox -e py38-test_galaxy_packages` work instead of using `TEST_PYTHON` ?

mr-c replied:
Perhaps if `python3.8` (or `python3.7` in my case) is on your PATH, sure. By adding this, `TEST_PYTHON=~/.pyenv/versions/3.7.12/bin/python tox -e test_galaxy_packages_for_pulsar` will work.

mr-c committed:
Makefile: add pyupgrade target

mr-c committed:
importlib_resources is only needed in Python 3.8

Add type hints + fixes

mr-c committed:
pin kombu to >= 5.3 for importlib-metadata fix (poetry had already picked up 5.3.x)

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -2,23 +2,29 @@
+    from importlib.resources import (  # type: ignore[attr-defined]
+        as_file,
+        files,
+        Package,
```
`Package` is deprecated in favour of `Anchor` since Python 3.12.
Should we import `Anchor` in a separate if/else, and revert the `python_version<'3.9'` condition on the `importlib-resources` dependency?

mr-c replied:
It is up to you, but isn't it nice to discard dependencies over time ? :-)

nsoranzo replied:
Of course, although I suspect we will have to change this again at some point. I've updated my suggestions accordingly.

nsoranzo commented on packages/util/setup.cfg:
```diff
@@ -34,8 +34,8 @@ include_package_data = True
-    docutils
-    importlib_resources
+    docutils!=0.17,!=0.17.1
+    importlib_resources;python_version<'3.9'
```
```suggestion
    importlib-resources;python_version<'3.9'
```

nsoranzo commented on packages/web_apps/setup.cfg:
```diff
@@ -42,12 +41,12 @@ install_requires =
     gunicorn
     gxformat2
-    importlib_resources
+    importlib_resources;python_version<'3.9'
```
```suggestion
    importlib-resources;python_version<'3.9'
```

nsoranzo commented on pyproject.toml:
```diff
@@ -68,10 +68,9 @@ gunicorn = "*"
 h5py = "*"
-importlib-metadata = "<5"  # Work around https://github.com/celery/kombu/issues/1600
-importlib-resources = "*"
+importlib-resources = { version = "*", python = "<3.9" }  # for importlib.{abc.Traversable,resources.{files, Package}}
```
```suggestion
importlib-resources = { version = "*", python = "<3.9" }  # for importlib.{abc.Traversable,resources.{Anchor, files}}
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -2,23 +2,29 @@
 
 
-def resource_path(package_or_requirement, resource_name):
+def resource_path(package_or_requirement: Package, resource_name: str) -> Traversable:
```
```suggestion
def resource_path(package_or_requirement: Anchor, resource_name: str) -> Traversable:
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -2,23 +2,29 @@
 
 
-def resource_string(package_or_requirement, resource_name) -> str:
+def resource_string(package_or_requirement: Package, resource_name: str) -> str:
```
```suggestion
def resource_string(package_or_requirement: Anchor, resource_name: str) -> str:
```

nsoranzo commented on test/unit/data/model/migrations/test_migrations.py:
```diff
@@ -1128,10 +1131,13 @@ def db_state6_gxy_state3_tsi_no_sam(url_factory, metadata_state6_gxy_state3_tsi_
+        traversable = resource_path("galaxy.model.migrations", "alembic.ini")
+        with as_file(traversable) as path:
+            config = Config(path)
+            path1, path2 = _get_paths_to_version_locations()
+            config.set_main_option("version_locations", f"{path1};{path2}")
+            # config.set_main_option() calls config.file_config() which memoizes
+            # the temporary file path from the as_file() contextmanager
```
These don't need to be in the `with` block?

mr-c replied:
They do need to be in the `with` block; that was the point of my comment :-) If they are outside the `with` block, then the file referenced by the `path` will already be deleted :-P

nsoranzo replied:
Sorry, I've overlooked your comment!

nsoranzo commented on lib/galaxy/util/__init__.py:
```diff
@@ -294,7 +295,7 @@ def unique_id(KEY_SIZE=128):
 
 
-def parse_xml(fname: StrPath, strip_whitespace=True, remove_comments=True) -> ElementTree:
+def parse_xml(fname: Union[StrPath, Traversable], strip_whitespace=True, remove_comments=True) -> ElementTree:
```
I am not sure that we want `Traversable` here, since for example we call `str(fname)` at line 306 below but the `Traversable` protocol doesn't have a `__str__()` method defined.

Maybe this is better:

```diff
diff --git a/test/unit/tool_util/data/test_tool_data_bundles.py b/test/unit/tool_util/data/test_tool_data_bundles.py
index 2e1a1f2717..9c6afddc57 100644
--- a/test/unit/tool_util/data/test_tool_data_bundles.py
+++ b/test/unit/tool_util/data/test_tool_data_bundles.py
@@ -15,7 +15,10 @@ from galaxy.util import (
     galaxy_directory,
     parse_xml,
 )
-from galaxy.util.resources import resource_path
+from galaxy.util.resources import (
+    as_file,
+    resource_path,
+)
 
 TOOLS_DIRECTORY = os.path.abspath(os.path.join(galaxy_directory(), "test/functional/tools/"))
 
@@ -51,8 +54,8 @@ def test_xml_parsing() -> None:
 
 
 def test_parsing_manual() -> None:
-    path = resource_path(__package__, "example_data_managers/manual.xml")
-    tree = parse_xml(path)
+    with as_file(resource_path(__package__, "example_data_managers/manual.xml")) as path:
+        tree = parse_xml(path)
     data_managers_el = tree.getroot()
     data_manager_el = data_managers_el.find("data_manager")
     description = convert_data_tables_xml(data_manager_el)
@@ -61,8 +64,8 @@ def test_parsing_manual() -> None:
 
 
 def test_parsing_mothur() -> None:
-    path = resource_path(__package__, "example_data_managers/mothur.xml")
-    tree = parse_xml(path)
+    with as_file(resource_path(__package__, "example_data_managers/mothur.xml")) as path:
+        tree = parse_xml(path)
     data_managers_el = tree.getroot()
     data_manager_el = data_managers_el.find("data_manager")
     description = convert_data_tables_xml(data_manager_el)
```

What do you think?

mr-c replied:
Good catch; I've applied this

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -1,24 +1,38 @@
+
+if sys.version_info >= (3, 12):
+    from importlib.resources import (
+        Anchor as PackageOrAnchor,
```
```suggestion
        Anchor,
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -1,24 +1,38 @@
+    from importlib.resources import (
+        as_file,
+        files,
+        Package as PackageOrAnchor,
```
```suggestion
        Package as Anchor,
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -1,24 +1,38 @@
+    from importlib_resources import (
+        as_file,
+        files,
+        Package as PackageOrAnchor,
```
```suggestion
        Package as Anchor,
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -1,24 +1,38 @@
+    from importlib_resources.abc import Traversable
+
+
+def resource_path(package_or_requirement: PackageOrAnchor, resource_name: str) -> Traversable:
```
```suggestion
def resource_path(package_or_requirement: Anchor, resource_name: str) -> Traversable:
```

nsoranzo commented on lib/galaxy/util/resources.py:
```diff
@@ -1,24 +1,38 @@
 
 
-def resource_string(package_or_requirement, resource_name) -> str:
+def resource_string(package_or_requirement: PackageOrAnchor, resource_name: str) -> str:
```
```suggestion
def resource_string(package_or_requirement: Anchor, resource_name: str) -> str:
```

nsoranzo commented on pyproject.toml:
```diff
@@ -137,6 +136,7 @@ galaxy-release-util = "*"
 httpx = "*"
 isort = "*"
 lxml = "!=4.2.2"
+lxml-stubs = "*"
```
Shouldn't this go to the `[tool.poetry.group.typecheck.dependencies]` section?

nsoranzo commented on lib/tool_shed/test/base/twillbrowser.py:
```diff
@@ -14,6 +14,8 @@
 
 tc.options["equiv_refresh_interval"] = 0
+# Resetting all repository metadata can take a really long time
+tc.timeout(120)
```
Should we go for 0?

mr-c replied:
😭 Seems so! (Go for it)

nsoranzo replied:
They finally pass with 240 s.
